core/src/main/java/org/bitcoinj/wallet/WalletProtobufSerializer.java

```diff
@@ -506,8 +514,15 @@ public Wallet readWallet(Network network, @Nullable WalletExtension[] extensions
         KeyChainGroup keyChainGroup;
         if (walletProto.hasEncryptionParameters()) {
             Protos.ScryptParameters encryptionParameters = walletProto.getEncryptionParameters();
-            final KeyCrypterScrypt keyCrypter = new KeyCrypterScrypt(encryptionParameters);
-            keyChainGroup = KeyChainGroup.fromProtobufEncrypted(network, walletProto.getKeyList(), keyCrypter, keyChainFactory);
+            if (walletProto.getEncryptionType() == EncryptionType.ENCRYPTED_SCRYPT_AES) {
+                final KeyCrypterScrypt keyCrypter = new KeyCrypterScrypt(encryptionParameters);
```

Comment:
> Is there a reason the KeyCrypterFactory isn't used here as well?

Line 518 could be changed to use the KeyCrypterFactory aswell which would require additional changes.

> do we really need it for line 521 – maybe a simple new KeyCrypterScrypt(…) would do as well?

The reason why we can’t do ‘new KeyCrypterScrypt()’ on line 521 is because the HWKeyCrypter (in bitcoin-wallet) that is created by the factory is a subclass of KeyCrypterScrypt. We chose to make it a subclass because it seemed to require the least amount of changes on bitcoinj side.

The new HWKeyCrypter does not use the provided encryption parameters, passwords or AesKeys. Ideally we’d implement a different interface that doesn’t require anything by default but that would require a lot more work and planning so we decided propose this lighter solution and see what comes back.

Another solution would be to have the HWKeyCrypter implement the KeyCrypter interface instead of extending the KeyCrypterScrypt. We did look into this solution early on but decided to go with the subclass solution. It wouldn’t make that much of a difference because the new HWKeyCrypter would still require unused parameters.